extensions/wikibase/src/org/openrefine/wikibase/schema/strategies/SnakOnlyStatementMerger.java

```diff
@@ -87,9 +89,9 @@ public boolean match(Snak existingSnak, Snak addedSnak) {
         // Deliberately only comparing the pids and not the siteIRIs to avoid spurious mismatches due to federation
         if (!existingSnak.getPropertyId().getId().equals(addedSnak.getPropertyId().getId())) {
             return false;
-        } else if (existingSnak instanceof NoValueSnak && addedSnak instanceof NoValueSnak) {
+        } else if (existingSnak instanceof NoValueSnakImpl && addedSnak instanceof NoValueSnakImpl) {
```

Comment:
I am surprised it is necessary actually! In which situations does it make a difference to check for `NoValueSnakImpl` instead of `NoValueSnak`? In general for this library the user (us) is expected to refer to the interfaces only, not their implementations, so this is a sign that something fishy is going on.

But yes, this code was indeed not used before, since we could not have an `addedSnak` that would be a `NoValueSnak`.

Reply:
In this case I never got the `NoValueSnak` variant to work for either `addedSnak` or `existingSnak`. Debugging indicated that `*Impl` instances were passed around.

However, I think the normal value snaks were passed around correctly.

Reply:
Ok! I guess the tests should help us get this to work. For the naming, the *Impl instances are also instances of the corresponding interfaces, so I really think this should not make any difference. Because the interfaces are meant to be the things we use here, I would prefer if we did not refer to the implementations.

Reply:
Do you have suggestions for good tests here, I'm drawn to the JSON-blob based ones but they aren't the easiest to follow on my end.

Reply:
I would write at least something in `WbStatementExprTest`. Just copy one of the existing test cases there (for instance `testEvaluate`) and replace one of the cell contents (defined in `setRow`) by one of your magic constants. Then, construct the expected value with the appropriate No/SomeValueSnak. That should be fairly straightforward.

Then, for the SnakOnlyStatementMerger, you could follow the same strategy: copy some existing test case, change the some/both of the involved statements being matched to have some no value / somevalue case, and specify the expected matching result.

Reply:
`WbStatementExprTest` doesn't seem to test anything other than `WbExpression` as the `evaluateTo` method is inherited from `WbExpressionTest`? At best I could catch the new exceptions.

Reply:
No, it does test `WbStatementExpr`. The evaluateExpr method can indeed work with any WbExpression, it is a utility that is used in many test classes. In this case it is used to test `WbStatementExpr` via a wrapper defined in the test class.

Reply:
We still need to change this to use the interfaces and not the implementations, just as a matter of good practice. Note that we are defining [FullyPropertySerializingNoValueSnak.java](https://github.com/OpenRefine/OpenRefine/pull/5360/files#diff-a9faad4d5d80e5c915fd5fe989e788ddf41d0205c9d7eb7d0734ea12d7d468d8) ourselves so we know those interfaces can have more implementations.

Reply:
Yeah sure, problem still is that changing to the interfaces break it.